### ISSUE TITLE
Odo source e2e test dotnet component push fails on Openshift 4.0 cluster

### ISSUE THREAD

**Issue by amitkrout** [redhat-developer/odo#1459]
[kind/bug]

## What versions of software are you using?
- Output of `odo version`:
```
$ odo version
odo v0.0.20 (3e634631)

Server: <URL>:6443
Kubernetes: v1.12.4+0f8e04e
```

## How did you run odo exactly?
$ git clone https://github.com/redhat-developer/odo && cd odo
$ make bin
$ ./odo project create odo-source
$ ./odo create dotnet dotnet-app --local tests/e2e/examples/source/dotnet/
$ ./odo push
```
OR
```
$ make test-source-e2e
## Actual behavior
Fails

## Expected behavior
Should pass

## Any logs, error output, etc?
```
$ ./odo create dotnet dotnet-app --local tests/e2e/examples/source/dotnet/
 ✓  Checking component
 ✓  Checking component version
 ✓  Initializing 'dotnet-app' component
 ✓  Creating component dotnet-app
 ✓  Component 'dotnet-app' was created and port 8080/TCP was opened
 ✓  Component 'dotnet-app' is now set as active component
To push source code to the component run 'odo push'
This may take a few moments to be ready
Amits-MacBook-Pro:odo amit$ ./odo push
Pushing changes to component: dotnet-app
 ✓  Waiting for component to start
 ✓  Copying files to component
 ◐  Building component ✗  Unable to build files
+ set -eo pipefail
+ '[' '!' -z /opt/app-root/src ']'
+ '[' /tmp '!=' /opt/app-root/src ']'
+ '[' /opt/app-root/src '!=' /opt/app-root/src ']'
+ '[' -f /tmp/src/.s2i/bin/assemble ']'
+ '[' -n /usr/libexec/s2i ']'
+ rm -rf /opt/app-root/src/.git
+ /usr/libexec/s2i/assemble
---> Installing application source...
Using SDK: 2.2.104
---> Restoring application dependencies...
  Restoring packages for /opt/app-root/src/asp-net-hello-world.csproj...
  Installing System.Threading.Overlapped 4.3.0.
  Installing System.Net.WebSockets 4.3.0.
  Installing System.Security.Principal 4.3.0.
  Installing System.Security.Claims 4.3.0.
  [...]
  Installing System.Globalization 4.3.0.
  Installing System.Net.Primitives 4.3.0.
  Generating MSBuild file /opt/app-root/src/obj/asp-net-hello-world.csproj.nuget.g.props.
  Generating MSBuild file /opt/app-root/src/obj/asp-net-hello-world.csproj.nuget.g.targets.
  Restore completed in 8.25 sec for /opt/app-root/src/asp-net-hello-world.csproj.
---> Publishing application...
Microsoft (R) Build Engine version 15.9.20.63311 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

/opt/rh/rh-dotnet22/root/usr/lib64/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): 
error NETSDK1005: Assets file '/opt/app-root/src/obj/project.assets.json' 
doesn't have a target for '.NETCoreApp,Version=v2.2'. 
Ensure that restore has run and that you have included 'netcoreapp2.2' 
in the TargetFrameworks for your project. [/opt/app-root/src/asp-net-hello-world.csproj]

 ✗  Building component
 ✗  command terminated with exit code 1
```

**Comment by amitkrout** [comment]
#1431

**Comment by amitkrout** [comment]
On 4.0 cluster 
```
$ odo catalog list componentsNAME        PROJECT       TAGS
dotnet      openshift     1.0,1.1,2.1,2.2,latest
httpd       openshift     2.4,latest
java        openshift     8,latest
nginx       openshift     1.10,1.12,1.8,latest
nodejs      openshift     10,6,8,8-RHOAR,latest
perl        openshift     5.24,5.26,latest
php         openshift     7.0,7.1,7.2,latest
python      openshift     2.7,3.5,3.6,latest
ruby        openshift     2.3,2.4,2.5,latest
wildfly     openshift     10.0,10.1,11.0,12.0,13.0,14.0,15.0,16.0,8.1,9.0,latest
```
On 3.11.0 cluster 
```
$ odo catalog list components
NAME           PROJECT            TAGS
dotnet         openshift          2.0,latest
httpd          openshift          2.4,latest
nginx          openshift          1.10,1.12,1.8,latest
nodejs         openshift          10,6,8,8-RHOAR,latest
perl           openshift          5.24,5.26,latest
php            openshift          7.0,7.1,latest
python         openshift          2.7,3.5,3.6,latest
ruby           openshift          2.3,2.4,2.5,latest
wildfly        openshift          10.0,10.1,11.0,12.0,13.0,8.1,9.0,latest
nodejs (*)     odo-1552548705     10,11,6,8,8-RHOAR,latest
```

**Comment by amitkrout** [comment]
As per the error resolution reported in the ```odo push```, by replacing ```<TargetFramework>netcoreapp2.0</TargetFramework>``` with ```<TargetFramework>netcoreapp2.2</TargetFramework>``` works on 4.0 cluster, but the same target framework tag for ```netcoreapp2.2``` doesnot work on 3.11 cluster.
> https://github.com/redhat-developer/odo/blob/master/tests/e2e/examples/source/dotnet/asp-net-hello-world.csproj#L4

**Comment by amitkrout** [comment]
However dotnet 2.0 is available in the image stream but the ```Tags``` is marked as ```hidden```
```
$ oc describe is -n openshift dotnet
[...]
2.0
  tagged from registry.centos.org/dotnet/dotnet-20-centos7:latest
    prefer registry pullthrough when referencing this tag

  RETIRED: Build and run .NET Core 2.0 applications on CentOS 7. For more information about using this builder image, including OpenShift considerations, see https://github.com/redhat-developer/s2i-dotnetcore/tree/master/2.0/build/README.md.
  Tags: hidden, builder, .net, dotnet, dotnetcore, rh-dotnet20
  Supports: dotnet:2.0, dotnet
  Example Repo: https://github.com/redhat-developer/s2i-dotnetcore-ex.git

  * registry.centos.org/dotnet/dotnet-20-centos7@sha256:0f8ba35b50a98c7980af6aa8b023f347e9b1d7f9195df7b99ddeb9afaa018a32
      3 minutes ago
    registry.redhat.io/dotnet/dotnet-20-rhel7@sha256:35cd9b0c2f596f0ffcb417946c97a3056005cda815d51705a49684f68c5b6521
      15 hours ago
[...]
```

**Comment by mohammedzee1000** [comment]
We should probably figure out what to do long term, but maybe for now we can skip dotnet against 4.0 cluster and get source tests on OpenShift CI but push for its removal from other CI's  last.

What you can do to achieve this
- setup an env OPENSHIFT_VERSION
- If this env is set and its value is 4.0, then skip dotnet tests

**Comment by amitkrout** [comment]
> We should probably figure out what to do long term, but maybe for now we can skip dotnet against 4.0 cluster and get source tests on OpenShift CI but push for its removal from other CI's last.
> 
> What you can do to achieve this
> 
>     * setup an env OPENSHIFT_VERSION
> 
>     * If this env is set and its value is 4.0, then skip dotnet tests
> 
>     * Edit tests in openshift/release to set `export OPENSHIFT_VERSION=4.0`

As a workaround even after removing the hidden tag from dotnet image stream - https://raw.githubusercontent.com/openshift/library/master/arch/x86_64/community/dotnet/imagestreams/dotnet-centos7.json and applying the same does not work. Also use of ```latest``` tag in the source test file could not help.

I am ok with suggestion. @kadel WDYT ?

**Comment by amitkrout** [comment]
^ Or lets not complicate the test. IMO this can be done just by changing ```netcoreapp2.0``` to ```netcoreapp2.2``` in [1]. I am proposing this because anyway 4.0 is going to be the default CI test environment and 3.11 will be used in some cases (mostly locally). 

So for our test we can create a known issue for source test on 3.11 cluster with a workaround. Later days we can fix it if needed. @mohammedzee1000 @kadel WDY ? 

[1] https://github.com/redhat-developer/odo/blob/master/tests/e2e/examples/source/dotnet/asp-net-hello-world.csproj#L4

**Comment by amitkrout** [comment]
#1469

**Comment by kadel** [comment]
> ^ Or lets not complicate the test. IMO this can be done just by changing `netcoreapp2.0` to `netcoreapp2.2` in [1]. I am proposing this because anyway 4.0 is going to be the default CI test environment and 3.11 will be used in some cases (mostly locally).

3.11 will be with us for a long time




> However dotnet 2.0 is available in the image stream but the `Tags` is marked as `hidden`

hidden shouldn't matter. It just hides it from `odo catalog list components` but you still can create component using it.

**Comment by kadel** [comment]
If this was solved in  https://github.com/redhat-developer/odo/pull/1469, it should be closed.

**Comment by amitkrout** [comment]
Fixed via #1469
